synapse/assets/storm/migrations/model-0.2.27.storm

```diff
@@ -0,0 +1,348 @@
+                    yield $n
+
+                    if $isarray {
+
+                        // We can't just [ :$prop-=$oldvalu :$prop+=$newvalu ] because the norm() function gets called
+                        // on the array type deep down in the AST. So, instead, we have to operate on the whole array.
+
+                        $list = $lib.copy(:$prop)
+                        while $list.has($oldvalu) {
+                            $list.remove($oldvalu)
+                        }
+
+                        if (not $list.has($newvalu)) {
+                            $list.append($newvalu)
+                        }
+
+                        $lib.model.migration.setNodePropValuNoNorm($node, $prop, $list)
+
+                    } else {
+                        try {
+                            [ -:$prop :$prop=$newvalu ]
+                        } catch ReadOnlyProp as exc {
+                            // The property is readonly so we can only delete it
+                            $lib.log.warning(`{$form}:{$prop} is readonly, cannot modify. Storing node: {$node.repr()}`)
+                            $removeNode($node, $view.iden, $view.layers.0.iden)
+                            | delnode
+
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    return()
+}
+
+// This is a separate function because I was getting weird bugs from scoping
+// issues. Mainly, the $view variable being overwritten because this code was
+// scoped to the whole script.
+function main() {
+
+    for $view in $lib.view.list(deporder=$lib.true) {
+
+        if (not $layers.has($view.layers.0.iden)) { continue }
+
+        view.exec $view.iden {
+
+            $layer = $lib.layer.get()
+
+            for $oldcpe in $layer.liftByProp(it:sec:cpe) {
+                $info = $lib.model.migration.s.itSecCpe_2_170_0_internal($oldcpe)
+
+                $success = ($info.status = "success")
+
+                if $success {
+                    // No primary property changes, nothing to do. Node has been fully migrated.
+                    if (not $info.valu) {
+                        continue
+                    }
+
+                    /*
+                     * At this point, we have a node that can be fixed but
+                     * needs to be migrated to a new node because the primary
+                     * property needs to be changed. We'll create a new
+                     * (correct) node, and copy everything from the old node.
+                     * Then we complete the migration by iterating through
+                     * all the views to fix the references.
+                     */
+
+                    $lib.log.debug(`Migrating invalid it:sec:cpe node: {$oldcpe.repr()} -> {$info.valu}`)
+                    [ it:sec:cpe=$info.valu ]
+                    $migrateCpeNode($oldcpe, $node)
+
+                    $lib.log.debug(`Deleting invalid it:sec:cpe node: {$oldcpe.repr()}`)
```

Comment:
This log output doesn't add any additional context ( got everything we need above )